### PR TITLE
[msbuild] Remove unnecessary calls to AppleSdkSettings.Init

### PR DESCRIPTION
**Pull request by wachaudh**
The MSBuild task `DetectSDKLocations` is called for every class library in a sln. The call to `AppleSdkSettings.Init()` within `DetectSDKLocations` is expensive and not required every time. Removing unnecessary calls to `AppleSdkSettings.Init()` significantly reduces the build times for a large project.

### CODE REVIEW

File: msbuild/Xamarin.iOS.Tasks.Core/Tasks/DetectSdkLocationsTaskBase.cs [hunk 1]
```diff
@@ -88,7 +88,9 @@ public abstract class DetectSdkLocationsTaskBase : Task
 
 		public override bool Execute ()
 		{
-			AppleSdkSettings.Init ();
+			if (!AppleSdkSettings.IsValid)
+				AppleSdkSettings.Init ();
```
Comment (jonathanpeppers) [inline msbuild/Xamarin.iOS.Tasks.Core/Tasks/DetectSdkLocationsTaskBase.cs]: I would probably log both cases:
```csharp
Log.LogMessage (MessageImportance.Low, "Calling AppleSdkSettings.Init");
```
or:
```csharp
Log.LogMessage (MessageImportance.Low, "Skipping AppleSdkSettings.Init, already valid.");
```

You should also add a test that checks a build with no changes skips the `Init` call.

Something like this one:
https://github.com/xamarin/xamarin-macios/blob/4544fb8fcd987e53790c5bda5cd251a015a518d7/msbuild/tests/Xamarin.iOS.Tasks.Tests/ProjectsTests/XamarinForms.cs#L25-L28

You should be able to assert the "Skipping..." log message was printed.

Comment (jonathanpeppers) [reply msbuild/Xamarin.iOS.Tasks.Core/Tasks/DetectSdkLocationsTaskBase.cs]: Also add some timings in the commit message that shows the savings for multiple projects. 👍

**Comment by jstedfast** [comment]
I like the idea of logging, but the PR looks good as-is.

**Comment by jonathanpeppers** [comment]
> This is the type of change that scares me. Trivial looking mucking with global state.

@chamons I'm having @wachaudh test what happens if you change the Xcode path in the IDE.

It might need to call `Init` again in that case.

**Comment by chamons** [comment]
Maybe also try xcode-select on the command line? I'm not sure which it respects w\o looking.

**Comment by jstedfast** [comment]
> > This is the type of change that scares me. Trivial looking mucking with global state.
> 
> @chamons I'm having @wachaudh test what happens if you change the Xcode path in the IDE.
> 
> It might need to call `Init` again in that case.

@jonathanpeppers it will have to call Init again. That said, doesn't the MSBuild process exit once the build is complete? Or does the IDE keep the MSBuild process alive, thereby making global state stick around for the lifetime of the IDE being open?

**Comment by jonathanpeppers** [comment]
> Or does the IDE keep the MSBuild process alive, thereby making global state stick around for the lifetime of the IDE being open?

@jstedfast yes, MSBuild stays around. We do something similar here for Xamarin.Android (checking java and javac version). We just needed to make sure it was invalidated if the path to java changed.

It used this feature of MSBuild instead of a `static` class, though:

https://github.com/xamarin/xamarin-android/blob/master/Documentation/guides/MSBuildBestPractices.md#caching-in-msbuild-tasks

**Comment by rolfbjarne** [comment]
> if you change the Xcode path in the IDE.

We already require the IDE to restart when changing Xcode.

This is what VSfM says where you change the Xcode path:

> "These preferences will take effect next time you start Visual Studio"

**Pull request closed without merging**